### ISSUE TITLE
Installer: factor out hard-coded references to dataverse-4.0.war

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#2165]
As @akio-sone pointed out on Monday, the string `dataverse-4.0.war` appears in several places in our code base.

Once we increment the version in `pom.xml` to "4.0.1" per http://guides.dataverse.org/en/latest/developers/making-releases.html we'll need to make sure these scripts still work:

```
murphy:dataverse pdurbin$ grep -r dataverse-4.0.war *
scripts/database/fedora/rebuild-and-test:cp target/dataverse-4.0.war $GLASSFISH_HOME/glassfish/domains/domain1/autodeploy
scripts/deploy/apitest.dataverse.org/rebuild:/usr/local/glassfish4/glassfish/bin/asadmin deploy /tmp/dataverse-4.0.war
scripts/installer/install:my $WARFILE_LOCATION = "../../target/dataverse-4.0.war";
scripts/installer/install:    $WARFILE_LOCATION = "dataverse-4.0.war";
scripts/installer/Makefile:DATAVERSE_WAR_FILE=${INSTALLER_ZIP_DIR}/dataverse-4.0.war
scripts/installer/Makefile:${DATAVERSE_WAR_FILE}:	../../target/dataverse-4.0.war
scripts/installer/Makefile:	/bin/cp ../../target/dataverse-4.0.war ${DATAVERSE_WAR_FILE}
scripts/installer/Makefile:../../target/dataverse-4.0.war:
scripts/installer/Makefile:	@echo "ERROR: can't find application .war file (../../target/dataverse-4.0.war)!"
scripts/installer/README.txt:for example, the war file (once built) can be found in ../../target/dataverse-4.0.war
scripts/installer/README.txt:So the script itself knows to look for these files in 2 places (for example, ../../target/dataverse-4.0.war and ./dataverse-4.0.war)
scripts/installer/README.txt:dataverse-4.0.war
scripts/vagrant/install-dataverse.sh:WAR=/dataverse/target/dataverse-4.0.war
```

**Comment by pdurbin** [comment]
@scolapasta I believe this work needs to happen as part of 4.0.1 so I'm at least moving it to the candidate milestone.

**Comment by pdurbin** [comment]
@landreev it sounds like you're fine with tackling this so I'm giving this ticket to you and setting it to 4.0.1.

At out meeting just now we seem to have reached a consensus to remove the version number from the war file. I think this will make things easier going forward. We'll all just deploy `dataverse.war` (developers and users).

p.s. Please don't worry about any hard coded references to `dataverse-4.0.war` in scripts having anything to do with vagrant or fedora or apitest. I'll fix them eventually.

**Comment by landreev** [comment]
Oh, this should be in QA.

**Comment by landreev** [comment]
Both the installer, and the Makefile that builds the installer zip bundle, have been cleared of all the hardcoded version numbers.

**Comment by akio-sone** [comment]
There are at least two issues within the modified installer of branch 4.0.2: the path setting of VersionNumber.properties lacks "main" portion and the regex that tries to match the version string should be "version.number" instead of "dataverse.version"

**Comment by landreev** [comment]
Thanks Akio - will fix shortly.

**Comment by landreev** [comment]
OK, fixed.

**Comment by landreev** [comment]
Thinking about it, the installer shouldn't be relying on the number in versionNumber.properties. 
Reading the pom.xml file should be more reliable - as the war file is guaranteed to contain the version number from the pom file in its name. 
The versionNumber.properties should be auto-generated from the pom file somehow - maybe?

**Comment by bencomp** [comment]
I'm running into the Vagrant related mentions currently.

@landreev agreed that this dependency shouldn't be there! On a related note: I've mentioned the [Maven Release plugin](http://maven.apache.org/maven-release/maven-release-plugin/) before to @pdurbin. It does cool stuff, but maybe not "generate a file that contains the version number" (or [does it?](http://maven.apache.org/maven-release/maven-release-plugin/examples/run-goals-before-commit.html)).

Is the VersionNumber.properties actually necessary?

**Comment by landreev** [comment]
That property file's main purpose is for the Application to know, at run time, what its own version number is. 
So yes, I'd say it is necessary - unless there is a cleaner/easier/more standard way. 
It's just that, ideally, it should be auto-generated from, and dependent on, the main pom file. 

What kind of problems are you running into? Is it just that the number in the properties file is different from the version number in the pom.xml in your branch? Just making them identical should resolve this.

**Comment by bencomp** [comment]
It seems that increasing the version in the pom.xml worked. Naturally this should be the source of all other version number references :)

**Comment by landreev** [comment]
Yes, sure, it should only be maintained in one place. We'll (or I will) change this setup accordingly.

I am confused though - I really thought I had already increased the number in pom.xml to 4.0.2 in that branch... But maybe I didn't. 
Could you please push your pom.xml into 4.0.2? (I'm working on a different branch right now). 
Thanks.

(and, as a reminder for everybody, pom.xml should be modified, first thing, every time you create a branch with a new version number! - as outlined in Phil's instruction)

**Comment by pdurbin** [comment]
> Could you please push your pom.xml into 4.0.2

@bencomp doesn't have push access to to repo but I just merged his pull request: #2259

**Comment by akio-sone** [comment]
If maven-antrun-plugin is used, the version of Dataverse can be given by its pom.xml.  There are at least two ways, one is the currently used properties file and the other is an automatically generated class file (this file must be on the .ignore list):         
```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-antrun-plugin</artifactId>
    <version>1.3</version>
    <executions>
    <!-- case 1 -->
        <execution>
            <id>properties-based case</id>
            <goals>
                <goal>run</goal>
            </goals>
            <phase>generate-sources</phase>
            <configuration>
                <tasks>

                    <delete file="${project.build.sourceDirectory}/project.properties" />
                    <touch  file="${project.build.sourceDirectory}/project.properties"
                mkdirs="true" />
                    <echo   file="${project.build.sourceDirectory}/project.properties"
                append="true">
                        version.number=${pom.version}
                    </echo>

                </tasks>
            </configuration>
        </execution>
        
    <!-- case 2 -->

        <execution>
            <id>class-based case</id>
            <goals>
                <goal>run</goal>
            </goals>
            <phase>generate-sources</phase>
            <configuration>
                <tasks>
                    <property name="src.dir" value="${project.build.sourceDirectory}" />
                    <property name="package.dir" value="edu/unc/odum/modeshape/webapp/testbed" />
                    <property name="package.name" value="edu.unc.odum.modeshape.webapp.testbed" />
                    <property name="buildtime" value="${maven.build.timestamp}" />

                    <delete file="${src.dir}/${package.dir}/Version.java" />
                    <echo file="${src.dir}/${package.dir}/Version.java" message="package ${package.name};${line.separator}" />
                    <echo file="${src.dir}/${package.dir}/Version.java" append="true" message="public final class Version {${line.separator}" />
                    <echo file="${src.dir}/${package.dir}/Version.java" append="true" message=" public static String VERSION=&quot;${pom.version}&quot;;${line.separator}" />
                    <echo file="${src.dir}/${package.dir}/Version.java" append="true" message=" public static String BUILD_TIME=&quot;${buildtime}&quot;;${line.separator}" />
                    <echo file="${src.dir}/${package.dir}/Version.java" append="true" message="}${line.separator}" />
                </tasks>

            </configuration>
        </execution>

    </executions>
</plugin>
```

**Comment by bencomp** [comment]
I like this autogeneration of the version number file. Can this be implemented?

**Comment by pdurbin** [comment]
> I like this autogeneration of the version number file. Can this be implemented?

@bencomp this ticket is in QA but please feel free to open a new issue describing what you're looking for.

**Comment by landreev** [comment]
OK, I have removed the VersionNumber.properties file. The version number is now defined in just one place - pom.xml. 
The application gets the version number from the pom properties file that Maven places into the war file. 
And the installer gets it directly from pom.xml, when in the "dev" mode. 
So... we may finally be done with this long-playing ticket.

**Comment by pdurbin** [comment]
> OK, I have removed the VersionNumber.properties file. The version number is now defined in just one place - pom.xml

@landreev I'm going to move out of QA because we're seeing "v. 4.UNKNOWN" on my laptop and "caught IOException trying to read the Manifest" which seems to be: https://github.com/IQSS/dataverse/blob/570d3ff34f6a6b235f05c376f6577caa5c641dd5/src/main/java/edu/harvard/iq/dataverse/util/SystemConfig.java#L138

**Comment by pdurbin** [comment]
> caught IOException trying to read the Manifest

Fixed in 605cb4d . Thanks @landreev . Passing to QA.

**Comment by kcondon** [comment]
OK, built installer and ran it successfully on beta, has correct version and build, closing.
